libs/parallelism/algorithms/include/hpx/parallel/algorithms/exclusive_scan.hpp

```diff
@@ -71,43 +353,45 @@ namespace hpx { namespace parallel { inline namespace v1 {
         }
 
         ///////////////////////////////////////////////////////////////////////
-        template <typename FwdIter2>
+        template <typename IterPair>
         struct exclusive_scan
-          : public detail::algorithm<exclusive_scan<FwdIter2>, FwdIter2>
+          : public detail::algorithm<exclusive_scan<IterPair>, IterPair>
         {
             exclusive_scan()
               : exclusive_scan::algorithm("exclusive_scan")
             {
             }
 
-            template <typename ExPolicy, typename InIter, typename OutIter,
-                typename T, typename Op>
-            static OutIter sequential(ExPolicy, InIter first, InIter last,
-                OutIter dest, T const& init, Op&& op)
+            template <typename ExPolicy, typename InIter, typename Sent,
+                typename OutIter, typename T, typename Op>
+            static util::in_out_result<InIter, OutIter> sequential(ExPolicy,
+                InIter first, Sent last, OutIter dest, T const& init, Op&& op)
             {
                 return sequential_exclusive_scan(
                     first, last, dest, init, std::forward<Op>(op));
             }
 
-            template <typename ExPolicy, typename FwdIter1, typename T,
-                typename Op>
+            template <typename ExPolicy, typename FwdIter1, typename Sent,
+                typename FwdIter2, typename T, typename Op>
             static typename util::detail::algorithm_result<ExPolicy,
-                FwdIter2>::type
-            parallel(ExPolicy&& policy, FwdIter1 first, FwdIter1 last,
+                util::in_out_result<FwdIter1, FwdIter2>>::type
+            parallel(ExPolicy&& policy, FwdIter1 first, Sent last,
                 FwdIter2 dest, T init, Op&& op)
             {
-                typedef util::detail::algorithm_result<ExPolicy, FwdIter2>
-                    result;
-                typedef hpx::util::zip_iterator<FwdIter1, FwdIter2>
-                    zip_iterator;
-                typedef typename std::iterator_traits<FwdIter1>::difference_type
-                    difference_type;
+                using result = util::detail::algorithm_result<ExPolicy,
+                    util::in_out_result<FwdIter1, FwdIter2>>;
+                using zip_iterator =
+                    hpx::util::zip_iterator<FwdIter1, FwdIter2>;
+                using difference_type =
+                    typename std::iterator_traits<FwdIter1>::difference_type;
 
                 if (first == last)
-                    return result::get(std::move(dest));
+                    return result::get(std::move(
+                        util::in_out_result<FwdIter1, FwdIter2>{first, dest}));
 
-                difference_type count = std::distance(first, last);
+                difference_type count = detail::distance(first, last);
 
+                FwdIter1 last_iter = detail::advance_to_sentinel(first, last);
```

Comment:
Nothing for this PR, but I would like you to think about how we could avoid having to go over the sequence twice, once to count the elements and once to get to the end of the sequence.